### PR TITLE
ECO-3539 [Braintree] Cannot use a payment_method_nonce more than once error

### PR DESCRIPTION
**Pull request by kraal-spryker**
Ticket: https://spryker.atlassian.net/browse/ECO-3539

### CODE REVIEW

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * @method \SprykerEco\Zed\Braintree\Business\BraintreeFacadeInterface getFacade()
+ */
+class BraintreePlaceOrderPreCheckPlugin extends AbstractPlugin implements CheckoutPlaceOrderPreCheckPluginInterface
+{
+    /**
+     * Specification:
+     * - Checks a condition before the order is placed. If the condition fails, an error is added to the response transfer and 'false' is returned.
+     * - Check could be passed (returns 'true') along with errors added to the checkout response.
+     * - Don't use this plugin to write to a DB
```
Comment (voitovtihran) [inline src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php:27]: Dot is missing in the end of the line.

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    protected function checkForErrors(
+        BraintreeTransactionResponseTransfer $braintreeTransactionResponseTransfer,
+        CheckoutResponseTransfer $checkoutResponseTransfer
+    ): bool {
+        if ($braintreeTransactionResponseTransfer->getIsSuccess()) {
+            return true;
+        }
+
+        $errorCode = $braintreeTransactionResponseTransfer->getCode() ?: 500;
+        $error = new CheckoutErrorTransfer();
```
Comment (voitovtihran) [inline src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php:77]: $checkoutErrorTransfer

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ * @method \SprykerEco\Zed\Braintree\Communication\BraintreeCommunicationFactory getFactory()
+ */
+class BraintreePlaceOrderPreCheckPlugin extends AbstractPlugin implements CheckoutPreConditionPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     * Specification:
+     * - Checks a condition before the order is placed. If the condition fails, an error is added to the response transfer and 'false' is returned.
+     * - Check could be passed (returns 'true') along with errors added to the checkout response.
+     * - Don't use this plugin to write to a DB.
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php:31]: This is specification from an interface. Here we want to have this concrete plugin specific specification.

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     * @param \Generated\Shared\Transfer\CheckoutResponseTransfer $checkoutResponseTransfer
+     *
+     * @return bool
+     */
+    public function checkCondition(
+        QuoteTransfer $quoteTransfer,
+        CheckoutResponseTransfer $checkoutResponseTransfer
+    ): bool {
+        $quoteTransfer->requirePayment();
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php:44]: The plugin method contains a business logic. That's not allowed according to our standards. We need to move all this logic behind facade.

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+use Spryker\Zed\Kernel\Communication\AbstractPlugin;
+use SprykerEco\Shared\Braintree\BraintreeConfig;
+
+/**
+ * @method \SprykerEco\Zed\Braintree\Business\BraintreeFacadeInterface getFacade()
+ * @method \SprykerEco\Zed\Braintree\BraintreeConfig getConfig()
+ * @method \SprykerEco\Zed\Braintree\Persistence\BraintreeQueryContainerInterface getQueryContainer()
+ * @method \SprykerEco\Zed\Braintree\Communication\BraintreeCommunicationFactory getFactory()
+ */
+class BraintreePlaceOrderPreCheckPlugin extends AbstractPlugin implements CheckoutPreConditionPluginInterface
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php:24]:
```suggestion
class BraintreeCheckoutPreConditionPlugin extends AbstractPlugin implements CheckoutPreConditionPluginInterface
```

**Review by aleksey-kotsuba — CHANGES_REQUESTED** [review]

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php (hunk 3 above)
Comment (ipartemk) [reply src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php]: done

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php (hunk 4 above)
Comment (ipartemk) [reply src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php]: done

File: src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php (hunk 5 above)
Comment (ipartemk) [reply src/SprykerEco/Zed/Braintree/Communication/Plugin/Checkout/BraintreePlaceOrderPreCheckPlugin.php]: done

**Review by aleksey-kotsuba — APPROVED** [review]

**Comment by romanhavrylko** [comment]
Re-create branch with correct name https://github.com/spryker-eco/braintree/pull/34

**Pull request closed without merging**